### DISCUSSION TITLE
[SOLVED] System tray issue under GNOME

### DISCUSSION THREAD

**Discussion by guihkx** [martinrotter/rssguard#350] [Ideas]
Hi, I don't know if you're too familiar with the GNOME desktop environment, so I figured I should give a little background first. In short, the entire concept of "system tray" (or "status icons") [is long gone from the GNOME desktop](https://blogs.gnome.org/aday/2017/08/31/status-icons-and-gnome/).

So you might be wondering, there are still A LOT of Linux programs using the system tray, so how does Ubuntu and other popular distros deal with this problem? Very simple, since you can develop extensions for GNOME Shell, these distros will ship with [this extension](https://github.com/ubuntu/gnome-shell-extension-appindicator/) pre-installed, which "emulates" the classic system tray.

Granted, that extension is not perfect, but most programs will work just fine with it. These extensions are very dynamic and will get automatically unloaded by GNOME Shell when, for example, you lock your screen. After you unlock it, then GNOME Shell loads them again.

Now onto the actual issue:

If you have enabled RSS Guard at system startup, you most certainly will experience these 3 issues:

1. After logging into your desktop, RSS Guard's main window will pop up. In my understanding, that's not the behavior you originally intended. In theory, RSS Guard was supposed to start minimized to the system tray, correct? Which brings us to issue number 2:
2. Because GNOME Shell extensions are written in Javascript, they might take a little longer to fully load when compared to some other programs, such as RSS Guard. And this problem can be aggravated if you have a large amount of GNOME Shell extensions installed. Therefore, because RSS Guard might start _sooner_ than the system tray extension itself, RSS Guard assumes the operating system doesn't have an usable system tray, which is why issue number 1 exists!
3. As I mentioned earlier, after you lock your screen, all GNOME Shell extensions will be unloaded (and this is not optional). So, even if RSS Guard had a working system tray icon prior to the locking, there will be no system tray again once you unlocks it. This issue seems to only affect Qt programs, though...

Now you might argue that issues 1 and 2 shouldn't happen in the first place, because you've added this [`X-GNOME-Autostart-Delay=15` option](https://github.com/martinrotter/rssguard/blob/2332da3359b20906b8c7b3c2e3253888d61d78e2/resources/desktop/com.github.rssguard.desktop.autostart#L9) to RSS Guard's autostart file. However, that option is _not_ recognized by vanilla GNOME, it is merely a [patch](https://salsa.debian.org/gnome-team/gnome-session/-/blob/1e57c0ee5a73ba2be64169c661c563dafdb1b567/debian/patches/ubuntu/support_autostart_delay.patch) used by Ubuntu/Debian. So, if you're running GNOME on Arch, for example, they will have no effect whatsoever.

### There is a solution!

Luckily, the great people working on [KeePassXC](https://github.com/keepassxreboot/keepassxc) have already worked on a workaround for this: https://github.com/keepassxreboot/keepassxc/pull/5948

In short, KeePassXC will attempt 5 times to get a working system tray (while waiting 5 seconds in between tries). If these 5 tries are exceeded, well, you're out of luck. But judging by the concept, this will at least fix the GNOME system tray issue.

The implementation seems pretty straightforward too (though I'm not a C++/Qt programmer, so I have no idea if implementing this on RSS Guard itself would be that easy), but since both projects share the same license, I thought this would be something worth mentioning.

Sorry for the wall of text. :)
Thanks for reading, and thanks for RSS Guard.

**Comment by martinrotter** [comment]
Hello.

I am relatively hardcore Linux user and I am familiar of tray icon situations and uprise of "app indicators". That said, I always avoided Gnome like a plague and stayed with KDE or MATE for long long time. I probably never ever tried RSS Guard on Gnome.

Looking at keepasxc's fix - they basically try to load tray icon several times until it succeedes, am I right? There is no other magic around, right?

One more thing, are you familiar with all options in
![image](https://user-images.githubusercontent.com/1255302/105852587-f60a8980-5fe4-11eb-8fdb-a444efb09d74.png)
? Particularly "Start application hidden". That could solve your issue number 1.

Point 3 is particularly worrying, and point 2 of course too. It is true, that I did not do any kind of extensive testing on Linux and even on Windows regarding to RSS Guard behavior when suspending system to RAM/disk and subsequently resuming it from suspended state.

**Comment by guihkx** [comment]
Thanks for the quick answer!

>Looking at keepasxc's fix - they basically try to load tray icon several times until it succeedes, am I right? There is no other magic around, right?

Yes, that's precisely it.

>"Start application hidden". That could solve your issue number 1.

I will test this option again later for my own sanity, but I'm pretty sure I already had it set and it had no effect. I don't know exactly what that option does, but thinking about it, where would RSS Guard "hide itself" if not in the system tray (which wouldn't been created yet)?

>It is true, that I did not do any kind of extensive testing on Linux and even on Windows regarding to RSS Guard behavior when suspending system to RAM/disk and subsequently resuming it from suspended state.

I think you misunderstood point 3. what I meant was that GNOME shell would kill the systray once the user locked the screen. And after the user has unlocked it, RSS Guard's tray icon will be gone (because the systray was killed by GNOME Shell), therefore requiring RSS Guard to "re-detect" the systray.

**Comment by guihkx** [comment]
I honestly think that by just implementing the same workaround KeePassXC did, all 3 issues would be solved (though I'm not _entirely_ sure about issue 3).

**Comment by guihkx** [comment]
I just learned hat there is an open pull request for the appindicator extension that fixes all 3 issues:

https://github.com/ubuntu/gnome-shell-extension-appindicator/pull/260

Unless a similar behavior exists in other desktop environments (other than GNOME 3), it seems that the KeePassXC workaround won't be necessary anymore. :)

**Reply by guihkx** [reply]
Hmm, it seems like this also happens on XFCE, judging by this bug report:

https://github.com/keepassxreboot/keepassxc/issues/5943

**Reply by martinrotter** [reply]
@guihkx It seems that there is D-Bus event/service `StatusNotifierWatcher` which triggers when tray becomes available, maybe I could just hook to the event and re-load tray icon if user has it enabled in settings?

**Reply by guihkx** [reply]
That sounds like it could work! I'd be happy to give it a try too.

**Reply by martinrotter** [reply]
Did some digging. Found out that Qt actually already somehow watches for that D-Bus signal:

https://code.qt.io/cgit/qt/qtbase.git/tree/src/platformsupport/themes/genericunix/dbusmenu/qdbusmenuconnection.cpp?h=v5.13.1#n70

I will try to do some testing on ubuntu/gnome to reproduce the bug and see the behavior myself.

**Reply by martinrotter** [reply]
@guihkx OK, I checked the situation and made some changes which themselves DO NOT fix the bug, but they will help RSS Guard to fully work when https://github.com/ubuntu/gnome-shell-extension-appindicator/pull/260 gets fixed.

After some thoughts I strongly believe that this is something that should be really really really fixed upstream.

**Reply by martinrotter** [reply]
Feel free to continue this discussion if there will be problems in RSS Guard even after this gets fixed upstream.

**Reply by guihkx** [reply]
>OK, I checked the situation and made some changes which themselves DO NOT fix the bug, but they will help RSS Guard to fully work when ubuntu/gnome-shell-extension-appindicator#260 gets fixed.

Hmm that's interesting. I'll tell you that right now I'm still on RSS Guard 3.8.4 and I have manually applied that patch to the appindicator extension, and I can assure you that RSS Guard's tray icon is always fully restored after I unlock my session.

The tray icon also shows up after I log in and RSS Guard opens *before* the appindicator extension itself.

Nonetheless, if you made changes that improved the situation, they're more than welcome too!

Thank you very much!

**Reply by martinrotter** [reply]
> 
> 
> > OK, I checked the situation and made some changes which themselves DO NOT fix the bug, but they will help RSS Guard to fully work when [ubuntu/gnome-shell-extension-appindicator#260](https://github.com/ubuntu/gnome-shell-extension-appindicator/pull/260) gets fixed.
> 
> Hmm that's interesting. I'll tell you that right now I'm still on RSS Guard 3.8.4 and I have manually applied that patch to the appindicator extension, and I can assure you that RSS Guard's tray icon is always fully restored after I unlock my session.
> 
> The tray icon also shows up after I log in and RSS Guard opens _before_ the appindicator extension itself.
> 
> Nonetheless, if you made changes that improved the situation, they're more than welcome too!
> 
> Thank you very much!

It could be that I did not actually manage to reproduce the bug fully, it was behaving weirdly, I just made sure that now tray icon C++ object is created even if user wants tray icons and there is no tray area, just to be sure they are there if Qt detects tray area later.

**Reply by guihkx** [reply]
Ah, gotcha.
